File: frontend/src/lib/schemas/constraint.ts
import { z } from 'zod';

export const constraintSchema = z.object({
  uid: z.string().uuid({ message: "uid must be a valid UUID" }),
  team_id: z.number(),
  year: z.string().regex(/^U([4-9]|1[0-9]|2[0-4])$/),
  start_time: z.string()
    .regex(/^([01]\d|2[0-3]):([0-5]\d)$/, "Must be in HH:MM format")
    .nullable()
    .optional(),
  length: z.number(),
  day_of_week: z.union([
    z.literal(0), z.literal(1), z.literal(2), z.literal(3), z.literal(4), z.literal(5), z.literal(6)
  ]).nullable().optional(),
  required_cost: z.union([
    z.literal(125), z.literal(250), z.literal(500), z.literal(1000)
  ]).nullable().optional(),
  field_id: z.number().nullable().optional(),
});

export type Constraint = z.infer<typeof constraintSchema>;

File: backend/models/constraint.py
from pydantic import BaseModel, Field
from typing import Optional, Literal
import re
from uuid import UUID

class Constraint(BaseModel):
    uid: UUID
    team_id: int
    year: str = Field(pattern=r'^U([4-9]|1[0-9]|2[0-4])$')
    start_time: Optional[str] = Field(None, pattern=r'^([01]\d|2[0-3]):([0-5]\d)$')  # HH:MM format
    length: int
    day_of_week: Optional[Literal[0, 1, 2, 3, 4, 5, 6]] = None
    required_cost: Optional[Literal[125, 250, 500, 1000]] = None
    field_id: Optional[int] = None


